lib/logstash/filters/dictionary/file.rb

```diff
@@ -45,7 +45,7 @@ def initialize(path, refresh_interval, exact, regex)
       initialize_for_file_type
       args = [@dictionary, rw_lock]
       if exact
-        @fetch_strategy = regex ? FetchStrategy::File::ExactRegex.new(*args) : FetchStrategy::File::ExactRegex.new(*args)
+        @fetch_strategy = regex ? FetchStrategy::File::ExactRegex.new(*args) : FetchStrategy::File::Exact.new(*args)
```

Comment:
This bit would likely read better as a `case` statement that selects the fetch strategy class; I have taken some liberty here, since we currently also select `RegexUnion` when both `regex` and `exact` are false, which looks like a bug to me:

~~~ ruby
      fetch_strategy_class = case
                             when exact && regex then FetchStrategy::File::ExactRegex
                             when exact         then FetchStrategy::File::Exact
                             when          regex then FetchStrategy::File::RegexUnion
                             else
                               fail('no fetch strategy found')
                             end
      @fetch_strategy = fetch_strategy_class.new(*args)
~~~

Reply:
Hmmm. Compared to the 3.1.0 code the current code looks very explanatory 😃 
```
      if @exact
        if @regex
          key = @dictionary.keys.detect{|k| source.match(Regexp.new(k))}
          if key
            event.set(@destination, lock_for_read { @dictionary[key] })
            matched = true
          end
        elsif @dictionary.include?(source)
          event.set(@destination, lock_for_read { @dictionary[source] })
          matched = true
        end
      else
        translation = lock_for_read { source.gsub(Regexp.union(@dictionary.keys), @dictionary) }
        
        if source != translation
          event.set(@destination, translation.force_encoding(Encoding::UTF_8))
          matched = true
        end
      end
```
I read this as if it is not exact, then use the RegexUnion mode.

Reviewer, Colin, and I discussed the klass.new approach but we were ambivalent and decided to leave it as is at the time.

The docs say:
> When `exact => true`, the translate filter will populate the destination field
with the exact contents of the dictionary value. When `exact => false`, the
filter will populate the destination field with the result of any existing
destination field's data, with the translated value substituted in-place.

and

> If you'd like to treat dictionary keys as regular expressions, set `regex => true`.
Note: this is activated only when `exact => true`.

I must admit, I find the Regex union mode quite obtuse but I suppose there could be cases where a field might have "sadly" and "sadness" and the dictionary has `"sad" => "glad"`.

I like your "boolean truth table" formatting of the case statement but I'm on the fence about it.

Calling in a 3rd tiebreaker @colinsurprenant WDYT?

Reply:
The docs do call out that the `regex` option only makes sense when `exact` is true, so what I though is a bug is in fact documented behaviour:

> Note: this is activated only when exact ⇒ true.